streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java

```diff
@@ -200,6 +201,7 @@ synchronized void evict() {
         }
         remove(eldest);
         cache.remove(eldest.key);
+        dirtyKeys.remove(eldest.key);
```

Comment:
I guess my main point is that, like Damian said, if we add the `dirtyKeys.remove` function we may hide some bugs that will make it even harder to detect in the future, so I'd rather bite the bullet and make it more detectible and fix it than adding this one line change. Other changes LGTM and we can merge as is.